cdc/entry/schema/snapshot.go

```diff
@@ -0,0 +1,1145 @@
+			return nil, nil
+		}
+		tbInfo := binlogInfo.TableInfo
+		if tbInfo == nil {
+			log.Warn("ignore a invalid DDL job", zap.Reflect("job", job))
+			return nil, nil
+		}
+		tableID := tbInfo.ID
+		table, ok := s.doPhysicalTableByID(tableID)
+		if !ok {
+			return nil, cerror.ErrSchemaStorageTableMiss.GenWithStackByArgs(job.TableID)
+		}
+		return table, nil
+	}
+}
+
+// NewSingleSnapshotFromMeta creates a new single schema snapshot from a tidb meta
+func NewSingleSnapshotFromMeta(meta *timeta.Meta, currentTs uint64, forceReplicate bool) (*Snapshot, error) {
+	// meta is nil only in unit tests
+	if meta == nil {
+		snap := NewEmptySnapshot(forceReplicate)
+		snap.currentTs = currentTs
+		return snap, nil
+	}
+	return NewSnapshotFromMeta(meta, currentTs, forceReplicate)
+}
+
+// Snapshot stores the source TiDB all schema information and it should be READ-ONLY!
+// If no special comments, all public methods are thread-safe.
+type Snapshot struct {
+	// map[versionedEntityName] -> int64
+	// The ID can be `-1` which means the table is deleted.
+	tableNameToID *btree.BTree
+
+	// map[versionedEntityName] -> int64
+	// The ID can be `-1` which means the table is deleted.
+	schemaNameToID *btree.BTree
+
+	// map[versionedID] -> *timodel.DBInfo
+	// The target can be `nil` which means the entity is deleted.
+	schemas *btree.BTree
+
+	// map[versionedID] -> *model.TableInfo
+	// The target can be `nil` which means the entity is deleted.
+	tables *btree.BTree
+
+	// map[versionedID] -> *model.TableInfo
+	partitions *btree.BTree
+
+	// map[versionedID] -> struct{}
+	truncatedTables *btree.BTree
+
+	// map[versionedID] -> struct{}
+	// Partitions and tables share ineligibleTables because their IDs won't conflict.
+	ineligibleTables *btree.BTree
+
+	// if forceReplicate is true, treat ineligible tables as eligible.
+	forceReplicate bool
+
+	currentTs uint64
+
+	rwlock *sync.RWMutex
+}
+
+// NewEmptySnapshot creates an empty schema snapshot.
+func NewEmptySnapshot(forceReplicate bool) *Snapshot {
+	return &Snapshot{
+		tableNameToID:    btree.New(16),
+		schemaNameToID:   btree.New(16),
+		schemas:          btree.New(16),
+		tables:           btree.New(16),
+		partitions:       btree.New(16),
+		truncatedTables:  btree.New(16),
+		ineligibleTables: btree.New(16),
+		forceReplicate:   forceReplicate,
+		currentTs:        0,
+		rwlock:           new(sync.RWMutex),
+	}
+}
+
+// NewSnapshotFromMeta creates a schema snapshot from meta.
+func NewSnapshotFromMeta(meta *timeta.Meta, currentTs uint64, forceReplicate bool) (*Snapshot, error) {
+	snap := NewEmptySnapshot(forceReplicate)
+	dbinfos, err := meta.ListDatabases()
+	if err != nil {
+		return nil, cerror.WrapError(cerror.ErrMetaListDatabases, err)
+	}
+	tag := negative(currentTs)
+
+	for _, dbinfo := range dbinfos {
+		vid := newVersionedID(dbinfo.ID, tag)
+		vid.target = unsafe.Pointer(dbinfo)
+		snap.schemas.ReplaceOrInsert(vid)
+
+		vname := newVersionedEntityName(-1, dbinfo.Name.O, tag)
```

Comment:
```suggestion
		vname := newVersionedEntityName(-1, dbinfo.Name.O, tag)
```
Please add some comment about the magic number `-1`.